### ISSUE TITLE
If node.main_image is not set, scan body and use the first image

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#3104]
Not every node has a main image (see example: https://publiclab.org/tag/mobius), but we could use the first image in the body text if one doesn't exist, by scanning body text and returning the first URL in the format:

`_____.png` OR with `.jpg` or `.jpeg` or `.gif`

The code to return a main image is here:

https://github.com/publiclab/plots2/blob/aa7678e9f937f3086619efa528de67070fc5e1cb/app/models/node.rb#L307-L317

**Comment by imacubsfan23** [comment]
Can I take this?

**Comment by jywarren** [comment]
Please, that'd be great!

**Comment by imacubsfan23** [comment]
How do I set the main_image if there isn't one yet? 
My guess would be `main_image = '<p><img src="/i/2.png" title="new_bg.png" alt="image description" /></p>', but that's not working for me.
Do I need to make a new image node?

**Comment by imacubsfan23** [comment]
@jywarren So I left a commit for what I have so far. It renders the first picture seen in the body of the note page. The only issue I'm having is making the picture clickable.

To make it clickable, I believe I would have to set `@node.main_image` to the image I have rendered now, but I don't know how to do that.

Also, what should we do about notes with no pictures in the body and no main image?

**Comment by jywarren** [comment]
Ah, do you have a PR I could look at? I see your commit but so I can leave inline comments and such. Thanks!

So, what if we solve this within the `.main_image` function instead of inside the view. So on line 310 here, we can check if there is still no image returned, then scan through the node body:

https://github.com/publiclab/plots2/blob/12743783ed07a058d4777da2f3c48873d6103aa3/app/models/node.rb#L307-L317

Also, I think we should probably write a couple tests for this, where we create a node in the text fixture data that includes an image, and demonstrate that `.main_image` can grab it out. While we're at it, we could try to demonstrate that it /won't/ grab, say, a pdf, or some other type. Including, I'd guess, external images that lack `https://`? What do you think?

Thanks!

**Comment by imacubsfan23** [comment]
PR here: #3171 

The pictures in the node body are only rendered, they are not Images. So, I'm not sure how to actually make it an Image in order to set .main_image to that image. Unless maybe they are images, but when I was returning `@node.images`, I wasn't getting anything back.

**Comment by jywarren** [comment]
Ah, i see! so normally we'd do `node.main_image.path(...)` inside a line
like this:

<img src="<%= @node.main_image.path(:large) %>" />

But here we are only getting a URL. That makes sense. Hmm. What if we at
least made a method like `node.scraped_image` to return the path. Then we
can have some logic in the views, but not the actual image extraction --
just the logic of whether to use `main_image` or `scraped_image` (unless
you can think of a better name). How's that?

**Comment by imacubsfan23** [comment]
Absolutely, I can do that. 

Do you still want me to write specific tests for this? I have 2 pages on my local build where the body is the same, but 1 has a main image and one doesn't. That's how I've been testing so far

**Comment by imacubsfan23** [comment]
Is this what you meant?

**Comment by jywarren** [comment]
Yep! Then we can unit test `scraped_image`. I think upon unit testing we may find we need to refine the scraper a bit... let's see!

**Comment by imacubsfan23** [comment]
That comment just made me realize how wrong my solution was. I'm gonna fix my code before I write any tests, I'll add another commit.

**Comment by jywarren** [comment]
actually if you want to write the tests first, let them fail, and then
refine your scraper, that's called Test Driven Development -- pretty neat
workflow! If you're interested, give it a try!
